File: python/100-Same-Tree.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# region Recursion


class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if p is None and q is None:
            return True
        if p is None or q is None:
            return False
        if p.val == q.val:
            return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)
        return False


# endregion Recursion

# region Iteration


class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        stack = []

        def check(p, q) -> bool:
            if p is None and q is None:
                return True
            if p is None or q is None:
                return False
            if p.val != q.val:
                return False
            return True

        stack.append((p, q))

        while stack:
            p, q = stack.pop()

            if not check(p, q):
                return False
            if p:
                stack.append((p.left, q.left))
                stack.append((p.right, q.right))
        return True


# endregion Iteration

File: c#/100-Same-Tree.cs
#region DFS
/// <summary>
/// LC 100: Same Tree
/// https://leetcode.com/problems/same-tree/
/// Approach: Recursive DFS
/// Time: O(n) - n is the number of nodes in the tree
/// Space: O(h) - h is the height of the tree
///  Best Case (Balanced Tree): O(log(n))
///  Worst Case (Skewed Tree): O(n)
/// </summary>
public class Solution {
    public bool IsSameTree(TreeNode p, TreeNode q) {
        if (p == null && q == null) return true;
        if (p == null || q == null) return false;

        return p.val == q.val &&
                IsSameTree(p.left, q.left) &&
                IsSameTree(p.right, q.right);

    }
}
#endregion